### ISSUE TITLE
Secure websockets receives data but fails to send

### ISSUE THREAD

**Issue by matthill** [zaphoyd/websocketpp#544]
I setup a wss client connection with a local Python websockets server.  I'm using C++11 rather than Boost.  I'm using a self-signed certificate and disabling certificate validation.

When I connect, the server sends a welcome message, and then my code immediately sends back an echo.  The welcome message is received, but then it fails when I send data back.  I get an error message (short read).  Here's the detailed log:

```
[2016-04-11 22:57:22] [devel] endpoint constructor
[2016-04-11 22:57:22] [devel] client constructor
[2016-04-11 22:57:22] [devel] asio::init_asio
[2016-04-11 22:57:22] [devel] set_message_handler
[2016-04-11 22:57:22] [devel] create_connection
[2016-04-11 22:57:22] [devel] asio con transport constructor
[2016-04-11 22:57:22] [devel] connection constructor
[2016-04-11 22:57:22] [devel] transport::asio::init
Initializing TLS handler
[2016-04-11 22:57:22] [devel] starting async DNS resolve for localhost:8000
You are secure: 1
Endpoint is listening: 0
I am the server: 0
[2016-04-11 22:57:22] [devel] Async DNS resolve successful. Results: 127.0.0.1:8000 
[2016-04-11 22:57:22] [devel] Starting async connect
[2016-04-11 22:57:22] [devel] asio handle_resolve_timeout timer cancelled
[2016-04-11 22:57:22] [devel] Async connect to 127.0.0.1:8000 successful.
[2016-04-11 22:57:22] [connect] Successful connection
[2016-04-11 22:57:22] [devel] connection start
[2016-04-11 22:57:22] [devel] asio connection init
[2016-04-11 22:57:22] [devel] asio connection handle pre_init
[2016-04-11 22:57:22] [devel] asio connection post_init
[2016-04-11 22:57:22] [devel] asio handle_connect_timeout timer cancelled
[2016-04-11 22:57:22] [devel] asio connection handle_post_init
[2016-04-11 22:57:22] [devel] connection handle_transport_init
[2016-04-11 22:57:22] [devel] connection send_http_request
[2016-04-11 22:57:22] [devel] Raw Handshake request:
GET / HTTP/1.1
Connection: Upgrade
Host: localhost:8000
Sec-WebSocket-Key: AAAAAAAAAAAAAAAAAAAAAA==
Sec-WebSocket-Version: 13
Upgrade: websocket
User-Agent: WebSocket++/0.7.0


[2016-04-11 22:57:22] [devel] asio post init timer cancelled
[2016-04-11 22:57:22] [devel] handle_send_http_request
[2016-04-11 22:57:22] [devel] asio async_read_at_least: 1
[2016-04-11 22:57:22] [devel] asio con handle_async_read
[2016-04-11 22:57:22] [devel] handle_read_http_response
[2016-04-11 22:57:22] [devel] Raw response: HTTP/1.1 101 Switching Protocols
Connection: Upgrade
Sec-WebSocket-Accept: ICX+Yqv66kxgM0FcWaLWlFLwTAI=
Upgrade: websocket


[2016-04-11 22:57:22] [connect] WebSocket Connection 127.0.0.1:8000 v-2 "WebSocket++/0.7.0" / 101
[2016-04-11 22:57:22] [devel] p = 0 bytes transferred = 0
[2016-04-11 22:57:22] [devel] asio async_read_at_least: 1
[2016-04-11 22:57:22] [devel] open handshake timer cancelled
[2016-04-11 22:57:22] [devel] asio con handle_async_read
[2016-04-11 22:57:22] [devel] p = 0 bytes transferred = 10
[2016-04-11 22:57:22] [devel] calling consume with 10 bytes
[2016-04-11 22:57:22] [devel] Processing Bytes: 82 08 57 65 6C 63 6F 6D 65 21 
[2016-04-11 22:57:22] [devel] bytes left after consume: 0
[2016-04-11 22:57:22] [devel] Complete message received. Dispatching
on_message called with hdl: 0x184c7c0 and message: Welcome!
[2016-04-11 22:57:22] [devel] connection send
[2016-04-11 22:57:22] [devel] write_push: message count: 1 buffer size: 41
[2016-04-11 22:57:22] [devel] asio async_read_at_least: 1
[2016-04-11 22:57:22] [devel] write_pop: message count: 0 buffer size: 0
[2016-04-11 22:57:22] [frame_header] Dispatching write containing 1 message(s) containing 6 header bytes and 41 payload bytes
[2016-04-11 22:57:22] [frame_header] Header Bytes: 
[0] (6) 82 A9 00 00 00 00 

[2016-04-11 22:57:22] [devel] connection handle_write_frame
[2016-04-11 22:57:22] [devel] asio con handle_async_read
[2016-04-11 22:57:22] [error] handle_read_frame error: asio.ssl:335544539 (short read)
[2016-04-11 22:57:22] [devel] connection terminate
[2016-04-11 22:57:22] [devel] asio connection async_shutdown
[2016-04-11 22:57:22] [info] asio async_shutdown error: asio.ssl:335544539 (short read)
[2016-04-11 22:57:22] [devel] connection handle_terminate
[2016-04-11 22:57:22] [devel] handle_terminate error: asio.ssl:335544539 (short read)
[2016-04-11 22:57:22] [disconnect] Disconnect close local:[1006,short read] remote:[1006]
[2016-04-11 22:57:22] [devel] asio socket shutdown timer cancelled
Done running!

```

**Comment by matthill** [comment]
Created a gist for the code here: https://gist.github.com/matthill/0bf049fa852d2f161a5ce41ee3e9511a

**Comment by zaphoyd** [comment]
tls short read is a generic error generally equivalent to unexpected EOF in non-TLS connections. This error can show up for hundreds of different reasons. The first thing I'd try and do is see if you can get any logging output out of the server that might explain why it is dropping the connection. Without information of that sort I can only guess.

A minor error in the program that may or may not have an effect: the use of `asio::ssl::context::tlsv1_client` context. This will hard code the TLS version to TLSv1.0. TLSv1.0 may be rejected by a security conscious server `asio::ssl::context::sslv23_client` is a more generic context that will choose the best version available (hopefully v1.2).

Other things I might check... Are you able to connect to this server with browser clients? Are you able to connect using this client to other WSS servers?

Finally, what OS/platform is this being built & run on. What version of OpenSSL is being used?

**Comment by matthill** [comment]
Thanks, Peter.  Just to close the loop on this, I believe you're right about the server.

I was using Python Tornado websockets server.  It did work successfully with a Python client.  It also worked with a browser client, but I needed to switch to a signed cert.  This was on Ubuntu 14.04 x64.

I switched websocketpp to connect to wss://websockets.org and it worked.  Then I switched my server to Autobahn over twisted, and it also worked.  So, I think I'll switch over to use that server.  

Thanks for a great library.  I looked at a couple other C/C++ libraries and I find your API to be the most intuitive.
